ipm: Add support for multi-channel IPM

carlocaione opened the pull request:
DNM before #38616 and #39035 are in.

One limitation of the current IPM API is that it is assuming that the hardware is only exporting one single channel through which the data can be sent or signalling can happen.
    
If the hardware supports multiple channels, the device must be instantiated (possibly in the DT) several times, one for each channel to be able to send data through multiple channels using the same hw peripheral. Also in the current API only one callback can be registered, that means that only one driver is controlling all the signalling happening on all the channels.
 
With this patch we introduce the possibility to send data through multiple channels and register a callback on different channels also when the IPM device has a single instance in the DT. The goal is giving the possibility to several drivers to send data through different channels and register a callback on a specific channel.
    
For sending data a new `channel` parameter has been added to the `ipm_send()` function, and the same parameter has been added to the `ipm_register_callback()` function.
    
Using this new parameter the driver can use `ipm_send()` to send data through a specific channel, and they can use `ipm_register_callback()` to register a callback on that channel, the callback will be called when new data will be received on that channel.
    
Two more functions are also added to the API: `ipm_max_channels_get()` to get the maximum number of channels and `ipm_configure_channel()` to configure the channel at the hardware level.
    
Three new DT helpers are also added:
- `DT_IPMS_CTLR_BY_NAME`
- `DT_IPMS_CHANNEL_BY_NAME`
- `IPM_DT_SPEC_GET`
    
to be able to easily retrieve device and TX / RX channels from DT.
    
So for example in the DT we can have:
 ``` 
        ipm: ipm@2a000 {
                compatible = "vnd,ipm";
                #mbox-cells = <1>;
                ...
        };

        ipm_zero: ipm@3a000 {
                compatible = "vnd,ipm";
                #mbox-cells = <0>;
                ...
        };
    
        ipm-user@0 {
                compatible = "some,ipm-user";
                mboxes = <&ipm 9>, <&ipm 8>;
                mbox-names =  "tx", "rx";
                ...
        };
        ipm-user@1 {
                compatible = "other,ipm-user";
                mboxes = <&ipm 10>, <&ipm 11>, <&ipm_zero>;
                mbox-names =  "tx", "rx", "single";
                ...
        };
 ```

`ipm-user@0` is using the channel `9` for TX and channel `8` for RX. It can register a callback that will be called when new data is received by the IPM device on channel `8`. At the same time `ipm-user@1` driver is sending data through channel `10`, and it is interested in all the messages received on the channel `11`, for this reason it will be registering a callback on this channel.
    
In the code we use the DT helpers to retrieve all the parameters, for example:
```   
        struct device *tx = DEVICE_DT_GET(DT_IPMS_CTLR_BY_NAME(..., tx));
        struct device *rx = DEVICE_DT_GET(DT_IPMS_CTLR_BY_NAME(..., rx));
    
        uint32_t tx_channel = DT_IPMS_CHANNEL_BY_NAME(..., tx));
        uint32_t tx_channel = DT_IPMS_CHANNEL_BY_NAME(..., rx));
``` 
or we can use:
```
        struct ipm_dt_spec ipm = IPM_DT_SPEC_GET(..., tx, rx);

        ipm_send_dt(ipm, 0, msg);
        ...
```
Tests are modified to test all the newly introduced features.

hubertmis commented on include/drivers/ipm.h:
```diff
@@ -87,12 +87,137 @@ typedef void (*ipm_register_callback_t)(const struct device *port,
+ *
+ * This type is sufficient to hold all the possible channels of the IPM peripheral
+ */
+typedef uint8_t ipm_channel_id_t;
```
Some hardware peripherals used to implement ipm driver can utilize 32-bit register to pass "channel id" between cores. It could be better to extend this type to `uint32_t`

carlocaione replied:
ACK

hubertmis commented on include/drivers/ipm.h:
```diff
@@ -221,6 +346,185 @@ static inline int z_impl_ipm_set_enabled(const struct device *ipmdev,
+ *	       finishes. If there is deferred processing on the remote side,
+ *	       or you would like to queue outgoing messages and wait on an
+ *	       event/semaphore, you can implement that in a high-level driver
+ * @param id Message identifier. Values are constrained by
+ *        @a ipm_max_data_size_get since many boards only allow for a
+ *        subset of bits in a 32-bit register to store the ID.
```
What's actually the difference between channel and id in this API? I believe both parameters are intended to use some hardware mechanism to distinguish separated "streams" of data through single hardware peripheral. Is there a need to keep both in the same function?

carlocaione replied:
I tried to explain my interpretation in the comment here https://github.com/zephyrproject-rtos/zephyr/pull/38523#issuecomment-919127028

hubertmis commented on include/drivers/ipm.h:
```diff
@@ -221,6 +346,185 @@ static inline int z_impl_ipm_set_enabled(const struct device *ipmdev,
+}
+
+/**
+ * @brief Enable interrupts and callbacks for inbound channels.
```
Are all channels disabled by default? If they are, should we add a function to disable them run-time?

Alternatively, is there a need to enable or disable channels? Isn't it enough to enable a channel when a callback is registered?

carlocaione replied:
I think there is a need if for example the hardware needs to enable / disable the interrupt associated to that channel.

hubertmis replied:
But there is no function right now to disable? Or have I missed it?

carlocaione replied:
In theory you use the `enable` parameter:
```
@param enable Set to 0 to disable and to nonzero to enable.
```

hubertmis replied:
Right. The name and first line of configuration is misleading. How about renaming to `ipm_configure_channel()`?

hubertmis commented on include/drivers/ipm.h:
```diff
@@ -221,6 +346,185 @@ static inline int z_impl_ipm_set_enabled(const struct device *ipmdev,
+ *
+ * @param ipmdev Driver instance pointer.
+ *
+ * @return Maximum possible number of supported channels.
```
I'm thinking if `IPM_MULTICHANNEL_DT_SUPPORT` is necessary. This function could return 1 if given hardware does not support multiple channels.

carlocaione replied:
Oh you are right, I'll try to implement that after having gathered more comments about this PR.

hubertmis reviewed (commented):
API of ipm driver is getting more confusing to me. I don't understand what's the difference between `id` and `channel` (apart from the fact you can assign separated callbacks to channels, but no to ids). There are multiple similar functions like `ipm_send()` and `ipm_send_channel()`. I'm not sure which of them should I use when I create a generic module using ipm driver.

I think we should unify ideas in this API and reduce number of redundant approaches.

carlocaione commented:
> I don't understand what's the difference between id and channel (apart from the fact you can assign separated callbacks to channels, but no to ids)

So, according to my interpretation on the original API, `id` is only a software interpretation of the `data` being sent through the only channel. For example you can have `id == SAMPLING_DATA, data == 0x31` if you are sending sampling data, or `id == TIMESTAMP, data == 0xdeadbeef` for timestamps.

On the contrary `channel` is an hardware feature and identify a physical connection. So for example you can have two channels, connected to two different remote procs and you can send to both sampling data or timestamps (if the hardware allows it).

Truth is that for many IPM peripherals (for example the `IPM NRF driver`), the `id` is basically not used because the messaging is done using one-bit line (only signalling), so you are basically sending `1` or `0` through a channel.

carlocaione commented:
I'll fix the `buildkite` errors after having received more feedbacks.

hubertmis commented:
> `id` is only a software interpretation of the `data` being sent

If `id` is only a software interpretation, I would move it from the drivers code to a higher layer. But it looks some of ipm drivers use hardware registers to pass id value between cores.

These concepts seem similar to me. I think we need clearly separate what is `channel` and what is `id`, or remove one of them.

carlocaione commented:
> These concepts seem similar to me. I think we need clearly separate what is `channel` and what is `id`, or remove one of them.

@dcpleung can you clarify?

dcpleung commented:
> @dcpleung can you clarify what the `id` is?

IIRC, some hardware has a special message type ID where it is used to raise different type of interrupts. Though, the API was defined too long ago that I don't remember what that is anymore.

However, I think we can group `id` and `data` into a struct to be passed instead of two separate arguments. Maybe we can rename `data` to `msg` to further clarify the usage.

IMHO, an overhaul of the API might be better going forward. For example, instead of adding APIs to support multi-channels, change the send function to incorporate channel number. There will be less confusion on which function to use in the future. If you are going this route, can I make a request to add a target CPU as send function argument? :)

carlocaione commented:
> > @dcpleung can you clarify what the `id` is?
> 
> IIRC, some hardware has a special message type ID where it is used to raise different type of interrupts. Though, the API was defined too long ago that I don't remember what that is anymore.
> 
> However, I think we can group `id` and `data` into a struct to be passed instead of two separate arguments. Maybe we can rename `data` to `msg` to further clarify the usage.

No objection, just a lot of drivers to fix :cry: 

> IMHO, an overhaul of the API might be better going forward. For example, instead of adding APIs to support multi-channels, change the send function to incorporate channel number. There will be less confusion on which function to use in the future.

Agree, that would still require fixing the current drivers.

> If you are going this route, can I make a request to add a target CPU as send function argument? :)

Uhm, isn't that information already embedded in the channel you use? In my use case each channel is already connecting the IPM device / hw to different CPUs, but that information is carried implicitly in the channel number, so the user needs just to select the correct channel to communicate with different CPUs. @dcpleung what was your idea about the target CPU? Can you make an example?

dcpleung commented:
> > > @dcpleung can you clarify what the `id` is?
> > 
> > 
> > IIRC, some hardware has a special message type ID where it is used to raise different type of interrupts. Though, the API was defined too long ago that I don't remember what that is anymore.
> > However, I think we can group `id` and `data` into a struct to be passed instead of two separate arguments. Maybe we can rename `data` to `msg` to further clarify the usage.
> 
> No objection, just a lot of drivers to fix 
> 
> > IMHO, an overhaul of the API might be better going forward. For example, instead of adding APIs to support multi-channels, change the send function to incorporate channel number. There will be less confusion on which function to use in the future.
> 
> Agree, that would still require fixing the current drivers.

Yeah... overhauling APIs is not an easy task. One good thing is that there are _only_ 7 drivers as of now...
* Consolidating `id` and `data` into a struct is more work as it requires changes to code.
* Adding a channel argument to existing to the send API is (probably) trivial. None of the current drivers will be using it so you can just add the argument and `ARG_UNUSED()`.

> > If you are going this route, can I make a request to add a target CPU as send function argument? :)
> 
> Uhm, isn't that information already embedded in the channel you use? In my use case each channel is already connecting the IPM device / hw to different CPUs, but that information is carried implicitly in the channel number, so the user needs just to select the correct channel to communicate with different CPUs. @dcpleung what was your idea about the target CPU? Can you make an example?

On the Intel audio DSP side, we need to send messages to particular CPU, and that's why I was thinking about the CPU argument. But as you said, "channel" can be used for this. So there is no need for an extra CPU argument.

carlocaione committed:
ipm: Consolidate id and data into a struct

The IPM API is pretty old and the actual meaning of the ID parameter is
lost in history. The IPM drivers are using data and ID in different
ways, sometimes interchangeably. In any case both the parameters are
used to pass data across a single channel.

Before introducing support for multiple channels, consolidate ID and
data into a "message" struct and fix all the drivers / users.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
ipm: Add support for multiple channels

One limitation of the current IPM API is that it is assuming that the
hardware is only exporting one single channel through which the data can
be sent or signalling can happen.

If the hardware supports multiple channels, the device must be
instantiated (possibly in the DT) several times, one for each channel to
be able to send data through multiple channels using the same hw
peripheral. Also in the current API only one callback can be registered,
that means that only one driver is controlling all the signalling
happening on all the channels.

With this patch we introduce the possibility to send data through
multiple channels and register a callback on different channels also
when the IPM device has a single instance in the DT. The goal is giving
the possibility to several drivers to send data through different
channels and register a callback on a specific channel.

For sending data a new 'channel' parameter has been added to the
ipm_send() function, and the same parameter has been added to the
ipm_register_callback() function.

Using this new parameter the driver can use ipm_send() to send data
through a specific channel, and they can use ipm_register_callback() to
register a callback on that channel, the callback will be called when
new data will be received on that channel.

Two more functions are also added to the API: ipm_max_channels_get() to
get the maximum number of channels and ipm_configure_channel() to
configure the channel at the hardware level.

Three new DT helpers are also added:
- IPM_DT_CTLR_BY_NAME
- IPM_DT_CHANNEL_BY_NAME
- IPM_DT_SPEC_GET

to be able to easily retrieve device and TX / RX channels from DT.

So for example in the DT we can have:

    ipm: ipm@2a000 {
            compatible = "vnd,ipm";
            #mbox-cells = <1>;
            ...
    };

    ipm_zero: ipm@3a000 {
            compatible = "vnd,ipm";
            #mbox-cells = <0>;
            ...
    };

    ipm-user@0 {
            compatible = "some,ipm-user";
            mboxes = <&ipm 9>, <&ipm 8>;
            mbox-names = "tx", "rx";
            ...
    };

    ipm-user@1 {
            compatible = "other,ipm-user";
            mboxes = <&ipm 10>, <&ipm 11>, <&ipm_zero>;
            mbox-names = "tx", "rx", "single";
            ...
    };

ipm-user@0 is using the channel '9' for TX and channel '8' for RX. It
can register a callback that will be called when new data is received by
the IPM device on channel '8'. At the same time ipm-user@1 driver is
sending data through channel '10', and it is interested in all the
messages received on the channel '11', for this reason it will be
registering a callback on this channel.

In the code we use the DT helpers to retrieve all the parameters, for
example:

    struct device *tx = DEVICE_DT_GET(IPM_DT_CTLR_BY_NAME(..., tx));
    struct device *rx = DEVICE_DT_GET(IPM_DT_CTLR_BY_NAME(..., rx));

    uint32_t tx_channel = IPM_DT_CHANNEL_BY_NAME(..., tx));
    uint32_t tx_channel = IPM_DT_CHANNEL_BY_NAME(..., rx));

or we can use:

    struct ipm_dt_spec ipm = IPM_DT_SPEC_GET(..., tx);

    ipm_send_dt(ipm, 0, msg);
    ...

Tests are modified to test all the newly introduced features.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

dcpleung commented on drivers/ipm/ipm_handlers.c:
```diff
@@ -8,13 +8,11 @@
-	Z_OOPS(Z_SYSCALL_MEMORY_READ(data, size));
-	return z_impl_ipm_send((const struct device *)dev, wait, id,
-			       (const void *)data, size);
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(msg->data, msg->size));
```
Probably need to check if `msg` itself can be read.

carlocaione replied:
done

hubertmis commented:
I'm still not sure what is ipm instance, channel, and id. Please confirm if my understanding described below is correct.

Ipm driver instance shall be created for each hardware peripheral capable to pass signals (and optionally associated data) between CPU clusters in the system. Each instance shall be associated with unique address space (real, not virtual). Each instance includes a number of channels (min 1, max UINT32_MAX). Each channel may target other CPU cluster, also multiple channels can target the same CPU cluster. While a message is sent through a selected channel it might be accompanied with optional `id` number, if given hardware supports such accompanying number. Allowed range of allowed `id` values for given ipm instance is hardware dependent. `id` value shall be within 0 - N range, such that N is hardware dependent number in range 0 - UINT32_MAX. N=0 is a special case if hardware does not support `id` accompanying signal sent through a channel.

Ipm driver user can determine number of supported channels and max value of `id` using functions available in the driver API.

carlocaione commented:
> Ipm driver instance shall be created for each hardware peripheral capable to pass signals (and optionally associated data) between CPU clusters in the system. Each instance shall be associated with unique address space (real, not virtual). Each instance includes a number of channels (min 1, max UINT32_MAX). Each channel may target other CPU cluster, also multiple channels can target the same CPU cluster. While a message is sent through a selected channel it might be accompanied with optional `id` number, if given hardware supports such accompanying number. Allowed range of allowed `id` values for given ipm instance is hardware dependent. 

Everything is correct.

> `id` value shall be within 0 - N range, such that N is hardware dependent number in range 0 - UINT32_MAX. N=0 is a special case if hardware does not support `id` accompanying signal sent through a channel.

I didn't attach any particular meaning to `ID` or `N`, but we can assume that `ID` is not supported if `N=0`, but that's not enforced by the API.

> Ipm driver user can determine number of supported channels and max value of `id` using functions available in the driver API.

Correct.

hubertmis commented:
Could we add something like this to the driver or API documentation to help new users of ipm driver understand how it is supposed to work?

carlocaione commented:
> Could we add something like this to the driver or API documentation to help new users of ipm driver understand how it is supposed to work?

@hubertmis I added a plot and a brief explanation (cherry picking a lot from your explanation :) in `ipm,h`. Let me know if that helps.

carlocaione commented on include/drivers/ipm.h:
```diff
@@ -18,6 +18,42 @@
+ *                                    |
+ *
+ * An IPM device is a peripheral capable of passing signals (or data) between
+ * CPUs and clusters in the system. Each IPM instance is usually providing one
```
fixed

carlocaione commented on include/drivers/ipm.h:
```diff
@@ -18,6 +18,42 @@
+ *
+ * An IPM device is a peripheral capable of passing signals (or data) between
+ * CPUs and clusters in the system. Each IPM instance is usually providing one
+ * ore more channels, each one targeting one other CPU cluster (multiple
```
ups, fixed.

galak committed:
devicetree: Add mailbox bindings

Add general mboxes, mbox-names to base.yaml to be utilized by any
clients that use mailboxes.

Additionally add mailbox-controller.yaml for common properties shared
by all mailbox controller devices.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
edtlib: Handle special casing for mbox

The association between "mboxes" property is with "#mbox-cells"
and "mbox-names" requires special case handing since its not
a simple "s" at the end like for other phandle style properties
(like "dmas", "clocks", etc..)

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

carlocaione commented:
Rebased on top of #38523 and started using the proper mailbox bindings.

carlocaione commented:
The dev-review meeting approved this stable API change. This is ready again for review.

carlescufi commented:
@nika-nordic @anangl @doki-nordic @masz-nordic could you please take a look?

dcpleung commented on include/drivers/ipm.h:
```diff
@@ -232,6 +367,58 @@ static inline int z_impl_ipm_set_enabled(const struct device *ipmdev,
+ * @retval 0       On success.
+ * @retval -EINVAL If it isn't an inbound channel.
+ */
+__syscall int ipm_configure_channel(const struct device *ipmdev, uint32_t channel,
+				    int enable);
```
Wonder if it would be better to pass a `uint32_t` instead of a simple `enable`. We can reserve, for example, the first 8-bit for common IPM configurations, and the rest would be up to individual drivers. For now, the common part would only have 1-bit of enabling and disabling the channel, and some bits are reserved for future expansions.

carlocaione replied:
That looks like a good idea and if we have to touch the API better doing all at once. I'll prepare the change.

carlocaione replied:
I changed `s/int enable/uint32_t conf`

carlocaione commented:
@mbolivar-nordic I think you also want to take a look at the DT helpers here.

masz-nordic commented on include/drivers/ipm.h:
```diff
@@ -87,12 +143,104 @@ typedef void (*ipm_register_callback_t)(const struct device *port,
+
+/**
+ * @typedef ipm_configure_channel_t
+ * @brief Configure (enable/disable) channel in the IPM device
```
`ipm_configure_channel()` does not mention dsiabling channels in its description. Perhaps this could be aligned.

carlocaione replied:
Fixed.

carlocaione commented:
Now that I got two ACKs I'm marking this DNM waiting for #39035 and #38616 to be merged.

carlocaione commented on include/drivers/ipm.h:
```diff
@@ -118,46 +266,44 @@ __subsystem struct ipm_driver_api {
- * @param data Pointer to the data sent in the message.
- * @param size Size of the data.
+ * @param channel Channel to use
+ * @param msg Message struct
```
Fixed

arnopo commented on include/drivers/ipm.h:
```diff
@@ -18,6 +18,42 @@
+ * +----------+                       |        +----------+
+ *                                    |
+ *
+ * An IPM device is a peripheral capable of passing signals (or data) between
```
i wonder is should not instead   `An IPM device is a peripheral capable of passing signals (and data depending on peripherals)`.
means I suppose that a mailbox at least provide signaling  service

arnopo commented on include/drivers/ipm.h:
```diff
@@ -69,7 +105,8 @@ typedef void (*ipm_callback_t)(const struct device *ipmdev, void *user_data,
  */
-typedef int (*ipm_send_t)(const struct device *ipmdev, int wait, struct ipm_msg *msg);
+typedef int (*ipm_send_t)(const struct device *ipmdev, int wait, uint32_t channel,
+			  struct ipm_msg *msg);
```
How do you ensure that the channel is the one declared in the DT for the driver usage?
Based on this API any driver could send a message on any channel of the hardware devide without control, right?

It seems to me that using a new handle would be safer. It could be a constant structure containing the ipmdev and the channel declared in the device tree, something like 
struct mbox_chan {
	const struct device *ipmdev;
	uint32_t channel;
	ipm_callback_t cb;
	void *user_data;
};
typedef int (*ipm_send_t)(const struct mbox_chan channel,  struct ipm_msg *msg); 
typedef void (*ipm_register_callback_t)(const struct mbox_chan channel,
					ipm_callback_t cb,
					void *user_data);
...

carlocaione replied:
> How do you ensure that the channel is the one declared in the DT for the driver usage?

The driver using the IPM API should retrieve the proper channel from the DT using the `DT_*` helpers on the `mbox` bindings. I can make that more explicit if needed.

> Based on this API any driver could send a message on any channel of the hardware devide without control, right?

Yes, but do we care? The user is in charge of the correctness of the DT for its own hardware from where the channel numbers used in this API are retrieved from. I don't think we should babysit the developers :)

> It seems to me that using a new handle would be safer.

I think this is an unnecessary complication TBH.

arnopo replied:
> > * `#ipm-cells = <0>`: one instance of the mailbox per channel ( seems already exist in Zephyr)
> > * `#ipm-cells > <1>`. For example, a second parameter can be used for bidirectional mailboxes to specify RX or TX.
> 
> Just for clarity: `#ipm-cells` should really be `#mbox-cells`, the commit message is wrong, I'll fix that.
> 
> The difference between `#mbox-cells = <0>` and `#mbox-cells = <1>` is that in the first case the IPM driver will be just discarding whatever is passed as `channel` parameter, in the second case the driver will take that into accont.
> 
> > In this case, how to handle a generic client code to be able to support different IPM backends? Using a struct handle would make APi more flexible according to the ` #ipm-cells`.
> 
> You mean when an IPM user device must use two different IPM devices at the same time with different `#mbox-cells` sizes? Is this a common case?

My point here is a use case with `#mbox-cells = <2>` 
In https://github.com/carlocaione/zephyr/blob/upstream_ipm/samples/subsys/ipc/openamp/remote/src/main.c, how to address a mailbox defined with ` #mbox-cells = <2>`?

```
 ipm0: ipm@2a000 {
     compatible = "vnd,ipm0";
     #mbox-cells = <2>;
     ...
 };

 ipm-user@0 {
     compatible = "some,ipm-user";
     mboxes = <&ipm0 1 0>, <&ipm0 1 1>;  /* second parameter defines the RX or TX  */
     mbox-names=  "tx", "rx";
 };
```
I can see such definitions in Linux kernel DT. It can never happen in Zephyr but if happens, this would probably break the API. Or does this be addressed using the ipm_configure_channel? In such case can code be platform agnostic?

> 
> Without considering for now the DT part, if `IPM0` is the device using `#ipm-cells = <0>` and `IPM1` the device using `#ipm-cells = <1>` the user must do something like:
> 
> ```
> ipm_send(IPM0, <whatever_channel_it_is_discarded_anyway>)
> ipm_send(IPM1, <tx_channel_retrieved_from_dt>)
> ```
Right  if #mbox-cells cannot be greater than 1 your API is fine.

> 
> The real problem here is how do you describe this situation in the DT. How do you use `mboxes` and `mbox-names` to reference two devices with different `#mbox-cells`?
> 
> ```
> ipm0: ipm@2a000 {
>     compatible = "vnd,ipm0";
>     #mbox-cells = <0>;
>     ...
> };
> 
> ipm1: ipm@2b000 {
>     compatible = "vnd,ipm1";
>     #mbox-cells = <1>;
>     ...
> };
> 
> ipm-user@0 {
>     compatible = "some,ipm-user";
>     mboxes = <&ipm1 9>, <&ipm1 8>;
>     mbox-names=  "tx", "rx";
>     ...
>    
>     what about ipm0 ???
> };
> ```
I don't know if this usecase exists.

carlocaione replied:
> My point here is a use case with `#mbox-cells = <2>` In https://github.com/carlocaione/zephyr/blob/upstream_ipm/samples/subsys/ipc/openamp/remote/src/main.c, how to address a mailbox defined with ` #mbox-cells = <2>`?

Oooh ok, probably I got now your point (hopefully :)

As long as the IPM device has one of the cells named `channel` in the binding / YAML file (see https://github.com/carlocaione/zephyr/blob/upstream_ipm/dts/bindings/test/vnd%2Cipm.yaml#L15) then the DT helper (see https://github.com/carlocaione/zephyr/blob/upstream_ipm/include/drivers/ipm.h#L233) is able to retrieve the correct TX / RX channel number.

So, back to your example:
> ```
>  ipm0: ipm@2a000 {
>      compatible = "vnd,ipm0";
>      #mbox-cells = <2>;
>      ...
>  };
> 
>  ipm-user@0 {
>      compatible = "some,ipm-user";
>      mboxes = <&ipm0 1 0>, <&ipm0 1 1>;  /* second parameter defines the RX or TX  */
>      mbox-names=  "tx", "rx";
>  };
> ```

In this case I expect the binding for `vnd,ipm0` being something like:
```
...
mbox-cells:
  - whatever_parameter
  - channel
```
So in your code you can still do:
```
unsigned int tx_channel = DT_IPMS_CHANNEL_BY_NAME(DT_NODELABEL(ipm0), tx); // this will be 0
unsigned int rx_channel = DT_IPMS_CHANNEL_BY_NAME(DT_NODELABEL(ipm0), rx); // this will be 1
```
And use `tx_channel` and `rx_channel` with the API.

Does this make sense?

I will anyway extend the API to better support `ipm_dt_spec` so that we can retrieve channel and device using only one macro and use that directly as you suggested.
